mkdocs/docs/config.md
# Config

Goiabada uses environment variables and/or executable flags to configure its behavior. 

If both an environment variable and a flag are set, the flag takes precedence.

---

`GOIABADA_AUTHSERVER_BASEURL` and `GOIABADA_ADMINCONSOLE_BASEURL`

Flag: `--authserver-baseurl` and `--adminconsole-baseurl`

Default: http://localhost:9090 (auth server) and http://localhost:9091 (admin console)

Description: The base URL of the application. This is used for external access into the application.

---

`GOIABADA_AUTHSERVER_INTERNALBASEURL` and `GOIABADA_ADMINCONSOLE_INTERNALBASEURL`

Flag: `--authserver-internalbaseurl` and `--adminconsole-internalbaseurl`

Default: (empty string)

Description: The internal base URL of the application. This is used for internal communication (for instance, when the `adminconsole` needs to call the `token` endpoint on the `authserver`)

---

`GOIABADA_AUTHSERVER_LISTEN_HOST_HTTPS` and `GOIABADA_ADMINCONSOLE_LISTEN_HOST_HTTPS`

Flag: `--authserver-listen-host-https` and `--adminconsole-listen-host-https`

Default: 0.0.0.0 (all network interfaces)

Description: The host to listen on for HTTPS access.

---

`GOIABADA_AUTHSERVER_LISTEN_PORT_HTTPS` and `GOIABADA_ADMINCONSOLE_LISTEN_PORT_HTTPS`

Flag: `--authserver-listen-port-https` and `--adminconsole-listen-port-https`

Default: 9443 (auth server) and 9444 (admin console)

Description: The port to listen on for HTTPS access.

---

`GOIABADA_AUTHSERVER_LISTEN_HOST_HTTP` and `GOIABADA_ADMINCONSOLE_LISTEN_HOST_HTTP`

Flag: `--authserver-listen-host-http` and `--adminconsole-listen-host-http`

Default: 0.0.0.0 (all network interfaces)

Description: The host to listen on for HTTP access.

---

`GOIABADA_AUTHSERVER_LISTEN_PORT_HTTP` and `GOIABADA_ADMINCONSOLE_LISTEN_PORT_HTTP`

Flag: `--authserver-listen-port-http` and `--adminconsole-listen-port-http`

Default: 9090 (auth server) and 9091 (admin console)

Description: The port to listen on for HTTP access.

---

`GOIABADA_AUTHSERVER_TRUST_PROXY_HEADERS` and `GOIABADA_ADMINCONSOLE_TRUST_PROXY_HEADERS`

Flag: `--authserver-trust-proxy-headers` and `--adminconsole-trust-proxy-headers`

Default: false

Description: If you're using a reverse proxy, you should trust the headers sent by it. The `True-Client-IP`, `X-Real-IP` or the `X-Forwarded-For` header will be used to get the client's IP address. If you're not using a reverse proxy, you should set those variables to `false`.

---

`GOIABADA_AUTHSERVER_SET_COOKIE_SECURE` and `GOIABADA_ADMINCONSOLE_SET_COOKIE_SECURE`

Flag: `--authserver-set-cookie-secure` and `--adminconsole-set-cookie-secure`

Default: false

Description: This should always be set to `true` in production, when using HTTPS. The only scenario where they should be `false` is when testing locally with HTTP only.

---

`GOIABADA_AUTHSERVER_LOG_HTTP_REQUESTS` and `GOIABADA_ADMINCONSOLE_LOG_HTTP_REQUESTS`

Flag: `--authserver-log-http-requests` and `--adminconsole-log-http-requests`

Default: false

Description: If `true`, log the HTTP requests to console.

---

`GOIABADA_AUTHSERVER_CERTFILE` and `GOIABADA_ADMINCONSOLE_CERTFILE`

Flag: `--authserver-certfile` and `--adminconsole-certfile`

Default: (empty string)

Description: Certificate file for HTTPS. If empty, TLS will not be enabled.

---

`GOIABADA_AUTHSERVER_KEYFILE` and `GOIABADA_ADMINCONSOLE_KEYFILE`

Flag: `--authserver-keyfile` and `--adminconsole-keyfile`

Default: (empty string)

Description: key file for HTTPS. If empty, TLS will not be enabled.

---

`GOIABADA_AUTHSERVER_LOG_SQL` and `GOIABADA_ADMINCONSOLE_LOG_SQL`

Flag: `--authserver-log-sql` and `--adminconsole-log-sql`

Default: false

Description: If `true`, log all SQL statements to console.

---

`GOIABADA_AUTHSERVER_AUDIT_LOGS_IN_CONSOLE` and `GOIABADA_ADMINCONSOLE_AUDIT_LOGS_IN_CONSOLE`

Flag: `--authserver-audit-logs-in-console` and `--adminconsole-audit-logs-in-console`

Default: false

Description: If `true`, log audit messages to console.

---

`GOIABADA_AUTHSERVER_STATICDIR` and `GOIABADA_ADMINCONSOLE_STATICDIR`

Flag: `--authserver-staticdir` and `--adminconsole-staticdir`

Default: (empty string)

Description: The directory where the static files are located. If empty, it uses the static files embedded into the binary. This config is useful when you want to make modifications to the static files and serve them from the filesystem (perhaps via a docker volume, when running in a container).

---

`GOIABADA_AUTHSERVER_TEMPLATEDIR` and `GOIABADA_ADMINCONSOLE_TEMPLATEDIR`

Flag: `--authserver-templatedir` and `--adminconsole-templatedir`

Default: (empty string)

Description: The directory where the HTML templates are located. If empty, it uses the HTML templates embedded into the binary. This config is useful when you want to make modifications to the templates and serve them from the filesystem (perhaps via a docker volume, when running in a container).

---

`GOIABADA_DB_TYPE`

Flag: `--db-type`

Default: sqlite

Description: Currently `mysql`, `postgres`, `mssql`, `sqlite` are supported.

---

`GOIABADA_DB_USERNAME`

Flag: `--db-username`

Default: root

Description: DB user's name.

---

`GOIABADA_DB_PASSWORD`

Flag: `--db-password`

Default: (empty string)

Description: DB user's password.

---

`GOIABADA_DB_HOST`

Flag: `--db-host`

Default: localhost

Description: DB server hostname.

---

`GOIABADA_DB_PORT`

Flag: `--db-port`

Default: 3306

Description: DB server TCP port.

---

`GOIABADA_DB_NAME`

Flag: `--db-name`

Default: goiabada

Description: Database (schema) name.

---

`GOIABADA_DB_DSN`

Flag: `--db-dsn`

Default: file::memory:?cache=shared

Description: DSN of the database. Only applicable when db type is `sqlite`. When using a file, don't forget to add `?_pragma=busy_timeout=5000&_pragma=journal_mode=WAL`. 

Example: `file:/home/john/goiabada.db?_pragma=busy_timeout=5000&_pragma=journal_mode=WAL`

---

`GOIABADA_ADMIN_EMAIL`

Flag: `--admin-email`

Default: admin

Description: The email address of the admin user (the first user created). This is only relevant when the application is started for the first time.

---

`GOIABADA_ADMIN_PASSWORD`

Flag: `--admin-password`

Default: changeme

Description: The password of the admin user (the first user created). This is only relevant when the application is started for the first time.

---

`GOIABADA_APPNAME`

Flag: `--appname`

Default: Goiabada

Description: The name of the application. This is used in the UI and is only relevant when the application is started for the first time (it can be changed in the `adminconsole` later).

---


mkdocs/docs/envvars.md
# Environment variables

Goiabada uses environment variables to configure the application. 

The following table lists the environment variables and their default values:

####General settings
| <div style="width:190px">Name</div> | Description | <div style="width:150px">Default value</div> |
|:-----|:----------|:----------------|
| `GOIABADA_APPNAME` | The name of the application | `Goiabada` |
| `GOIABADA_ADMIN_EMAIL` | The email address of the admin user (the first user created) | `admin@example.com` |
| `GOIABADA_ADMIN_PASSWORD` | The password of the admin user (the first user created) | `changeme` |

####HTTP listener settings
| <div style="width:300px">Name</div> | Description | Default value |
|:-----|:----------|:----------------|
| `GOIABADA_KEYFILE` | PKCS8 key file for https.<br/>If empty, TLS will not be enabled. | empty |
| `GOIABADA_CERTFILE` | Certificate file for https.<br/>If empty, TLS will not be enabled. | empty |
| `GOIABADA_HOST` | Server's hostname.<br/> The empty string will make it listen on all network interfaces. | `localhost` if not in a container, or empty string if in a container. |
| `GOIABADA_PORT` | Server's TCP port.<br/>If empty and TLS enabled: `8443`, otherwise `8080`. | `8080` (http) or `8443` (https) |
| `GOIABADA_BASEURL` | Server's external URL.<br/>If empty, calculated from TLS enabled state, `GOIABADA_HOST` and `GOIABADA_PORT` | `http://localhost:8080` |
| `GOIABADA_ISSUER` | Value of `iss` field in the generated JWT tokens.<br/>If empty, equals to the `GOIABADA_BASEURL` | `http://localhost:8080` |
| `GOIABADA_STATICDIR` | The directory where the static files are located.<br/>If empty, uses the static files embedded into the binary. | empty |
| `GOIABADA_TEMPLATEDIR` | The directory where the HTML templates are located.<br/>If empty, uses the HTML templates embedded into the binary. | empty |
| `GOIABADA_ISBEHINDAREVERSEPROXY` | If you want to use a reverse proxy in front of Goiabada, set this to `true` | `false` |
| `GOIABADA_RATELIMITER_ENABLED` | An HTTP rate limiter is available to prevent brute force attacks. It's enabled by default. <br/>Some users prefer to apply an HTTP rate limiter from an external service like Cloudflare. If that's you, set this to `false`. | `true` |
| `GOIABADA_RATELIMITER_MAXREQUESTS` | The maximum number of requests allowed per time window.<br />Only relevant if the http rate limiter is enabled. | `50` |
| `GOIABADA_RATELIMITER_WINDOWSIZEINSECONDS` | The rate limiter window size in seconds.<br />Only relevant if the http rate limiter is enabled. | `10` |

####Database settings

| <div style="width:190px">Name</div> | Description | <div style="width:220px">Deafult value</div> |
|:-----|:----------|:----------------|
| `GOIABADA_DB_TYPE` | Currently `mysql` and `sqlite` are supported.<br/>For backward compatibility, the default value is `mysql`, but if `GOIABADA_DB_HOST` isn't defined, then the default is `sqlite`. | `mysql` |
| `GOIABADA_DB_HOST` | DB server hostname. | `localhost` |
| `GOIABADA_DB_PORT` | DB server TCP port. | `3306` |
| `GOIABADA_DB_USERNAME` | DB user's name. | `root` |
| `GOIABADA_DB_PASSWORD` | DB user's password. | empty |
| `GOIABADA_DB_DBNAME` | Database (schema) name. | `goiabada` |
| `GOIABADA_DB_DSN` | DSN of the database. Only applicable when db type is `sqlite`.<br /><br />When using a file, don't forget to add `?_pragma=busy_timeout=5000&_pragma=journal_mode=WAL` (see example on the right).  | `file::memory:?cache=shared`<br /><br />or<br /><br />`file:/home/john/goiabada.db?_pragma=busy_timeout=5000&_pragma=journal_mode=WAL` |

####Log settings
| <div style="width:320px">Name</div> | Description | Default value |
|:-----|:----------|:----------------|
| `GOIABADA_LOGGER_ROUTER_HTTPREQUESTS_ENABLED` | If `true`, log the HTTP requests. | `false` |
| `GOIABADA_AUDITING_CONSOLELOG_ENABLED` | If `true`, log audit messages to console. | `false` |
| `GOIABADA_LOGGER_GORM_TRACEALL` | If `true`, log all SQL statements to console. | `false` |

When starting Goiabada without any environment variable set, it will listen on `http://localhost:8080` and will use an in-memory SQLite database. 

The admin email and password will be `admin@example.com` and `changeme`. All changes will be lost upon restart. If you want a permanent test environment, specify the `GOIABADA_DB_DSN` = `file:./goiabada.db` environment variable.

src/core/config/config.go
package config

import (
	"flag"
	"os"
	"strconv"
	"strings"
	"sync"
)

type ServerConfig struct {
	BaseURL            string
	InternalBaseURL    string
	ListenHostHttps    string
	ListenPortHttps    int
	ListenHostHttp     string
	ListenPortHttp     int
	TrustProxyHeaders  bool
	SetCookieSecure    bool
	LogHttpRequests    bool
	CertFile           string
	KeyFile            string
	LogSQL             bool
	AuditLogsInConsole bool
	StaticDir          string
	TemplateDir        string
}

type DatabaseConfig struct {
	Type     string
	Username string
	Password string
	Host     string
	Port     int
	Name     string
	DSN      string
}

type Config struct {
	AuthServer    ServerConfig
	AdminConsole  ServerConfig
	Database      DatabaseConfig
	AdminEmail    string
	AdminPassword string
	AppName       string
}

var (
	cfg          Config
	activeConfig *ServerConfig
	once         sync.Once
)

// Init initializes the configuration and sets the active server
func Init(server string) {
	once.Do(load)
	setActiveServer(server)
}

// setActiveServer sets the active server configuration
func setActiveServer(server string) {
	switch server {
	case "AuthServer":
		activeConfig = &cfg.AuthServer
	case "AdminConsole":
		activeConfig = &cfg.AdminConsole
	default:
		panic("Invalid active server configuration specified")
	}
}

func Get() *ServerConfig {
	return activeConfig
}

func GetAuthServer() *ServerConfig {
	return &cfg.AuthServer
}

func GetAdminConsole() *ServerConfig {
	return &cfg.AdminConsole
}

func GetDatabase() *DatabaseConfig {
	return &cfg.Database
}

func GetAdminEmail() string {
	return cfg.AdminEmail
}

func GetAdminPassword() string {
	return cfg.AdminPassword
}

func GetAppName() string {
	return cfg.AppName
}

func load() {
	cfg = Config{
		AuthServer: ServerConfig{
			BaseURL:            getEnv("GOIABADA_AUTHSERVER_BASEURL", "http://localhost:9090"),
			InternalBaseURL:    getEnv("GOIABADA_AUTHSERVER_INTERNALBASEURL", ""),
			ListenHostHttps:    getEnv("GOIABADA_AUTHSERVER_LISTEN_HOST_HTTPS", "0.0.0.0"),
			ListenPortHttps:    getEnvAsInt("GOIABADA_AUTHSERVER_LISTEN_PORT_HTTPS", 9443),
			ListenHostHttp:     getEnv("GOIABADA_AUTHSERVER_LISTEN_HOST_HTTP", "0.0.0.0"),
			ListenPortHttp:     getEnvAsInt("GOIABADA_AUTHSERVER_LISTEN_PORT_HTTP", 9090),
			TrustProxyHeaders:  getEnvAsBool("GOIABADA_AUTHSERVER_TRUST_PROXY_HEADERS"),
			SetCookieSecure:    getEnvAsBool("GOIABADA_AUTHSERVER_SET_COOKIE_SECURE"),
			LogHttpRequests:    getEnvAsBool("GOIABADA_AUTHSERVER_LOG_HTTP_REQUESTS"),
			CertFile:           getEnv("GOIABADA_AUTHSERVER_CERTFILE", ""),
			KeyFile:            getEnv("GOIABADA_AUTHSERVER_KEYFILE", ""),
			LogSQL:             getEnvAsBool("GOIABADA_AUTHSERVER_LOG_SQL"),
			AuditLogsInConsole: getEnvAsBool("GOIABADA_AUTHSERVER_AUDIT_LOGS_IN_CONSOLE"),
			StaticDir:          getEnv("GOIABADA_AUTHSERVER_STATICDIR", ""),
			TemplateDir:        getEnv("GOIABADA_AUTHSERVER_TEMPLATEDIR", ""),
		},
		AdminConsole: ServerConfig{
			BaseURL:            getEnv("GOIABADA_ADMINCONSOLE_BASEURL", "http://localhost:9091"),
			InternalBaseURL:    getEnv("GOIABADA_ADMINCONSOLE_INTERNALBASEURL", ""),
			ListenHostHttps:    getEnv("GOIABADA_ADMINCONSOLE_LISTEN_HOST_HTTPS", "0.0.0.0"),
			ListenPortHttps:    getEnvAsInt("GOIABADA_ADMINCONSOLE_LISTEN_PORT_HTTPS", 9444),
			ListenHostHttp:     getEnv("GOIABADA_ADMINCONSOLE_LISTEN_HOST_HTTP", "0.0.0.0"),
			ListenPortHttp:     getEnvAsInt("GOIABADA_ADMINCONSOLE_LISTEN_PORT_HTTP", 9091),
			TrustProxyHeaders:  getEnvAsBool("GOIABADA_ADMINCONSOLE_TRUST_PROXY_HEADERS"),
			SetCookieSecure:    getEnvAsBool("GOIABADA_ADMINCONSOLE_SET_COOKIE_SECURE"),
			LogHttpRequests:    getEnvAsBool("GOIABADA_ADMINCONSOLE_LOG_HTTP_REQUESTS"),
			CertFile:           getEnv("GOIABADA_ADMINCONSOLE_CERTFILE", ""),
			KeyFile:            getEnv("GOIABADA_ADMINCONSOLE_KEYFILE", ""),
			LogSQL:             getEnvAsBool("GOIABADA_ADMINCONSOLE_LOG_SQL"),
			AuditLogsInConsole: getEnvAsBool("GOIABADA_ADMINCONSOLE_AUDIT_LOGS_IN_CONSOLE"),
			StaticDir:          getEnv("GOIABADA_ADMINCONSOLE_STATICDIR", ""),
			TemplateDir:        getEnv("GOIABADA_ADMINCONSOLE_TEMPLATEDIR", ""),
		},
		Database: DatabaseConfig{
			Type:     getEnv("GOIABADA_DB_TYPE", "sqlite"),
			Username: getEnv("GOIABADA_DB_USERNAME", "root"),
			Password: getEnv("GOIABADA_DB_PASSWORD", ""),
			Host:     getEnv("GOIABADA_DB_HOST", "localhost"),
			Port:     getEnvAsInt("GOIABADA_DB_PORT", 3306),
			Name:     getEnv("GOIABADA_DB_NAME", "goiabada"),
			DSN:      getEnv("GOIABADA_DB_DSN", "file::memory:?cache=shared"),
		},
		AdminEmail:    getEnv("GOIABADA_ADMIN_EMAIL", "admin"),
		AdminPassword: getEnv("GOIABADA_ADMIN_PASSWORD", "changeme"),
		AppName:       getEnv("GOIABADA_APPNAME", "Goiabada"),
	}

	// Auth server
	flag.StringVar(&cfg.AuthServer.BaseURL, "authserver-baseurl", cfg.AuthServer.BaseURL, "Goiabada auth server base URL")
	flag.StringVar(&cfg.AuthServer.InternalBaseURL, "authserver-internalbaseurl", cfg.AuthServer.InternalBaseURL, "Goiabada auth server internal base URL")
	flag.StringVar(&cfg.AuthServer.ListenHostHttps, "authserver-listen-host-https", cfg.AuthServer.ListenHostHttps, "Auth server https host")
	flag.IntVar(&cfg.AuthServer.ListenPortHttps, "authserver-listen-port-https", cfg.AuthServer.ListenPortHttps, "Auth server https port")
	flag.StringVar(&cfg.AuthServer.ListenHostHttp, "authserver-listen-host-http", cfg.AuthServer.ListenHostHttp, "Auth server http host")
	flag.IntVar(&cfg.AuthServer.ListenPortHttp, "authserver-listen-port-http", cfg.AuthServer.ListenPortHttp, "Auth server http port")
	flag.BoolVar(&cfg.AuthServer.TrustProxyHeaders, "authserver-trust-proxy-headers", cfg.AuthServer.TrustProxyHeaders, "Trust HTTP headers from reverse proxy in Auth server? (True-Client-IP, X-Real-IP or the X-Forwarded-For headers)")
	flag.BoolVar(&cfg.AuthServer.SetCookieSecure, "authserver-set-cookie-secure", cfg.AuthServer.SetCookieSecure, "Set secure flag on cookies for auth server")
	flag.BoolVar(&cfg.AuthServer.LogHttpRequests, "authserver-log-http-requests", cfg.AuthServer.LogHttpRequests, "Log HTTP requests for auth server")
	flag.StringVar(&cfg.AuthServer.CertFile, "authserver-certfile", cfg.AuthServer.CertFile, "Certificate file for HTTPS (auth server)")
	flag.StringVar(&cfg.AuthServer.KeyFile, "authserver-keyfile", cfg.AuthServer.KeyFile, "Key file for HTTPS (auth server)")
	flag.BoolVar(&cfg.AuthServer.LogSQL, "authserver-log-sql", cfg.AuthServer.LogSQL, "Log SQL queries for auth server")
	flag.BoolVar(&cfg.AuthServer.AuditLogsInConsole, "authserver-audit-logs-in-console", cfg.AuthServer.AuditLogsInConsole, "Enable audit logs in console output for auth server")
	flag.StringVar(&cfg.AuthServer.StaticDir, "authserver-staticdir", cfg.AuthServer.StaticDir, "Static files directory for auth server")
	flag.StringVar(&cfg.AuthServer.TemplateDir, "authserver-templatedir", cfg.AuthServer.TemplateDir, "Template files directory for auth server")

	// Admin console
	flag.StringVar(&cfg.AdminConsole.BaseURL, "adminconsole-baseurl", cfg.AdminConsole.BaseURL, "Goiabada admin console base URL")
	flag.StringVar(&cfg.AdminConsole.InternalBaseURL, "adminconsole-internalbaseurl", cfg.AdminConsole.InternalBaseURL, "Goiabada admin console internal base URL")
	flag.StringVar(&cfg.AdminConsole.ListenHostHttps, "adminconsole-listen-host-https", cfg.AdminConsole.ListenHostHttps, "Admin console https host")
	flag.IntVar(&cfg.AdminConsole.ListenPortHttps, "adminconsole-listen-port-https", cfg.AdminConsole.ListenPortHttps, "Admin console https port")
	flag.StringVar(&cfg.AdminConsole.ListenHostHttp, "adminconsole-listen-host-http", cfg.AdminConsole.ListenHostHttp, "Admin console http host")
	flag.IntVar(&cfg.AdminConsole.ListenPortHttp, "adminconsole-listen-port-http", cfg.AdminConsole.ListenPortHttp, "Admin console http port")
	flag.BoolVar(&cfg.AdminConsole.TrustProxyHeaders, "adminconsole-trust-proxy-headers", cfg.AdminConsole.TrustProxyHeaders, "Trust HTTP headers from reverse proxy in Admin console? (True-Client-IP, X-Real-IP or the X-Forwarded-For headers)")
	flag.BoolVar(&cfg.AdminConsole.SetCookieSecure, "adminconsole-set-cookie-secure", cfg.AdminConsole.SetCookieSecure, "Set secure flag on cookies for admin console")
	flag.BoolVar(&cfg.AdminConsole.LogHttpRequests, "adminconsole-log-http-requests", cfg.AdminConsole.LogHttpRequests, "Log HTTP requests for admin console")
	flag.StringVar(&cfg.AdminConsole.CertFile, "adminconsole-certfile", cfg.AdminConsole.CertFile, "Certificate file for HTTPS (admin console)")
	flag.StringVar(&cfg.AdminConsole.KeyFile, "adminconsole-keyfile", cfg.AdminConsole.KeyFile, "Key file for HTTPS (admin console)")
	flag.BoolVar(&cfg.AdminConsole.LogSQL, "adminconsole-log-sql", cfg.AdminConsole.LogSQL, "Log SQL queries for admin console")
	flag.BoolVar(&cfg.AdminConsole.AuditLogsInConsole, "adminconsole-audit-logs-in-console", cfg.AdminConsole.AuditLogsInConsole, "Enable audit logs in console output for admin console")
	flag.StringVar(&cfg.AdminConsole.StaticDir, "adminconsole-staticdir", cfg.AdminConsole.StaticDir, "Static files directory for admin console")
	flag.StringVar(&cfg.AdminConsole.TemplateDir, "adminconsole-templatedir", cfg.AdminConsole.TemplateDir, "Template files directory for admin console")

	// Database
	flag.StringVar(&cfg.Database.Type, "db-type", cfg.Database.Type, "Database type. Options: mysql, sqlite")
	flag.StringVar(&cfg.Database.Username, "db-username", cfg.Database.Username, "Database username")
	flag.StringVar(&cfg.Database.Password, "db-password", cfg.Database.Password, "Database password")
	flag.StringVar(&cfg.Database.Host, "db-host", cfg.Database.Host, "Database host")
	flag.IntVar(&cfg.Database.Port, "db-port", cfg.Database.Port, "Database port")
	flag.StringVar(&cfg.Database.Name, "db-name", cfg.Database.Name, "Database name")
	flag.StringVar(&cfg.Database.DSN, "db-dsn", cfg.Database.DSN, "Database DSN (only for sqlite)")

	// Initial setup
	flag.StringVar(&cfg.AdminEmail, "admin-email", cfg.AdminEmail, "Default admin email")
	flag.StringVar(&cfg.AdminPassword, "admin-password", cfg.AdminPassword, "Default admin password")
	flag.StringVar(&cfg.AppName, "appname", cfg.AppName, "Default app name")

	flag.Parse()
}

func getEnv(key string, defaultVal string) string {
	if value, exists := os.LookupEnv(key); exists {
		return strings.TrimSpace(value)
	}
	return strings.TrimSpace(defaultVal)
}

func getEnvAsInt(key string, defaultVal int) int {
	valueStr := getEnv(key, "")
	if value, err := strconv.Atoi(strings.TrimSpace(valueStr)); err == nil {
		return value
	}
	return defaultVal
}

func getEnvAsBool(key string) bool {
	valueStr := getEnv(key, "")
	if value, err := strconv.ParseBool(strings.TrimSpace(valueStr)); err == nil {
		return value
	}
	return false
}
